### PR TITLE
bugfix/13768-gantt-using-keys

### LINKED ISSUE #13768: Gantt- unable to create the chart using the keys
#### Expected behaviour
Create the chart by setting the `keys` and putting data into an array.

#### Actual behaviour
The chart is not creating.

#### Live demo with steps to reproduce
https://jsfiddle.net/BlackLabel/uz6Lf74y/1/

#### Product version
Highcharts Gantt JS v8.1.2 (2020-06-16)

#### Affected browser(s)
All

### PR DESCRIPTION
**Pull request by Basts**
Fixed #13768, unable to create a Gantt series with the `keys` option.

### CODE REVIEW

File: ts/Core/Axis/TreeGridAxis.ts [hunk 1]
```diff
@@ -479,15 +486,25 @@ namespace TreeGridAxis {
 
                     // Update yData now that we have calculated the y values
                     axis.series.forEach(function (series: Highcharts.Series): void {
-                        var data = (series.options.data || []).map(function (
+                        var axisData = (series.options.data || []).map(function (
                             d: Highcharts.PointOptionsType
                         ): Highcharts.PointOptionsType {
-                            return isObject(d, true) ? merge(d) : d;
+
```
Comment (TorsteinHonsi) [inline ts/Core/Axis/TreeGridAxis.ts:492]: Where did returning `merge(d)` go? The original function made sure to return a copy of `d` if `d` was an object...

File: ts/Core/Axis/TreeGridAxis.ts [hunk 2]
```diff
@@ -479,15 +487,25 @@ namespace TreeGridAxis {
...
-                        var data = (series.options.data || []).map(function (
+                        var axisData = (series.options.data || []).map(function (
                             d: Highcharts.PointOptionsType
                         ): Highcharts.PointOptionsType {
+
+                            if (series.options.keys && series.options.keys.length) {
+                                // Get the axisData from the data array used to
+                                // build the treeGrid where has been modified
+                                data.forEach(function (point: Highcharts.GanttPointOptions): void {
+                                    if ((d as any).includes(point.x) && (d as any).includes(point.x2)) {
```
Comment (pawelfus) [inline ts/Core/Axis/TreeGridAxis.ts:498]: Two things here:
- `includes` is not supported by IE at all
- what if `d: PointOptionsType` is object type? Or in general, not an array?

**Review by pawelfus — CHANGES_REQUESTED** [review]

File: ts/Core/Axis/TreeGridAxis.ts [hunk 3]
```diff
@@ -479,15 +487,25 @@ namespace TreeGridAxis {
 
                     // Update yData now that we have calculated the y values
                     axis.series.forEach(function (series: Highcharts.Series): void {
-                        var data = (series.options.data || []).map(function (
+                        var axisData = (series.options.data || []).map(function (
                             d: Highcharts.PointOptionsType
                         ): Highcharts.PointOptionsType {
+
+                            if (H.isArray(d) && series.options.keys && series.options.keys.length) {
```
Comment (pawelfus) [inline ts/Core/Axis/TreeGridAxis.ts:494]: Don't use `H.isArray`, but instead import `isArray` from utils.

File: ts/Core/Axis/GridAxis.ts [hunk 4]
```diff
@@ -664,13 +664,19 @@ class GridAxis {
                     ): boolean {
                         return (p as any)[axis.isXAxis ? 'x' : 'y'] === value;
                     });
+                let pointCopy;
 
+                if (point && series.is('gantt') && !(point as any).initialData) {
```
Comment (pawelfus) [inline ts/Core/Axis/GridAxis.ts:669]: I might be missing it, but what's this `initialData` property?

File: ts/Series/GanttSeries.ts [hunk 5]
```diff
@@ -352,14 +350,14 @@ seriesType<Highcharts.GanttSeries>('gantt', 'xrange'
             this: Highcharts.GanttPoint,
             options: Highcharts.GanttPointOptions,
             x: number
-        ): Point {
+        ): Highcharts.GanttPointOptions {
             var point = this,
-                retVal = merge(options) as any;
+                retVal = options;
```
Comment (pawelfus) [inline ts/Series/GanttSeries.ts:355]: This seems to be an unnecessary assignment?

File: ts/Gantt/Pathfinder.ts [hunk 6]
```diff
@@ -1081,6 +1081,11 @@ Pathfinder.prototype = {
         chart.series.forEach(function (series: Highcharts.Series): void {
             if (series.visible && !series.options.isInternal) {
                 series.points.forEach(function (point: Point): void {
+                    // For Gantt series the connect could be
+                    // defined as a dependency
+                    if (point.options && (point.options as Highcharts.GanttPointOptions).dependency) {
+                        (point.options as any).connect = (point.options as Highcharts.GanttPointOptions).dependency;
```
Comment (pawelfus) [inline ts/Gantt/Pathfinder.ts:1087]: Quite a lot castings here (to `any` and `GanttPointOptions`). Can we somehow improve this?

**Review by pawelfus — CHANGES_REQUESTED** [review]

**Review by pawelfus — APPROVED** [review]
Nothing more from me, thanks! :)

File: ts/Series/GanttSeries.ts [hunk 7]
```diff
@@ -352,15 +357,14 @@ seriesType<Highcharts.GanttSeries>('gantt', 'xrange'
...
             options: Highcharts.GanttPointOptions,
             x: number
-        ): Point {
-            var point = this,
-                retVal = merge(options) as any;
+        ): Highcharts.GanttPointOptions {
+            var point = this;
 
-            H.seriesTypes.gantt.prototype.setGanttPointAliases(retVal);
+            (options as any) = parent.prototype.pointClass.prototype.applyOptions
```
Comment (bre1470) [inline ts/Series/GanttSeries.ts:363]: `options` needs to be of type `DeepPartial<Highcharts.GanttPointOptions>`

**Review by bre1470 — CHANGES_REQUESTED** [review]
The `any` cast has to be removed.

File: ts/Series/GanttSeries.ts (hunk 7 above)
Comment (bre1470) [reply ts/Series/GanttSeries.ts:363]: Maybe use a new property instead of trying to replace the input argument.

**Review by bre1470 — APPROVED** [review]
Thanks! In type sense it is now okay.

It would be nice, if `setGanttPointAliases(ganttPoint as Highcharts.GanttPointOptions) ` can be written as `setGanttPointAliases(ganttPoint.options)`, but I guess some property changes get not reflected in the options property.

**Review by TorsteinHonsi — APPROVED** [review]

**Pull request merged**